pageserver/src/tenant/pricing.rs

```diff
@@ -0,0 +1,458 @@
+            let next_gc_cutoff = std::cmp::min(gc_info.horizon_cutoff, gc_info.pitr_cutoff);
+
+            let maybe_cutoff = if next_gc_cutoff > timeline.get_ancestor_lsn() {
+                // only include these if they are after branching point; otherwise we would end up
+                // with duplicate updates before the actual branching.
+                Some((next_gc_cutoff, LsnKind::GcCutOff))
+            } else {
+                None
+            };
+
+            // this assumes there are no other lsns than the branchpoints
+            let lsns = gc_info
+                .retain_lsns
+                .iter()
+                .inspect(|&&x| {
+                    trace!(
+                        timeline_id=%timeline.timeline_id,
+                        "retained lsn: {x:?}, is_before_ancestor_lsn={}",
+                        x < timeline.get_ancestor_lsn()
+                    )
+                })
+                .filter(|&&lsn| lsn > timeline.get_ancestor_lsn())
+                .cloned()
+                .map(|lsn| (lsn, LsnKind::BranchPoint))
+                .chain(maybe_cutoff)
+                .collect::<Vec<_>>();
+
+            (
+                lsns,
+                gc_info.horizon_cutoff,
+                gc_info.pitr_cutoff,
+                next_gc_cutoff,
+            )
+        };
+
+        // update this to have a retention_period later for the pricing_model
+        // it is currently unsure what this should be.
+        if let Some(cutoff_distance) = last_record_lsn.checked_sub(next_gc_cutoff) {
+            match min_max_cutoff_distance.as_mut() {
+                Some((min, max)) => {
+                    *min = std::cmp::min(*min, cutoff_distance);
+                    *max = std::cmp::max(*max, cutoff_distance);
+                }
+                _ => {
+                    min_max_cutoff_distance = Some((cutoff_distance, cutoff_distance));
+                }
+            }
+        }
+
+        // all timelines branch from something, because it might be impossible to pinpoint
+        // which is the pricing models "default" branch.
+        updates.push(Update {
+            lsn: timeline.get_ancestor_lsn(),
+            command: Command::BranchFrom(timeline.get_ancestor_timeline_id()),
+            timeline_id: timeline.timeline_id,
+        });
+
+        for (lsn, _kind) in &interesting_lsns {
+            if let Some(size) = logical_size_cache.get(&(timeline.timeline_id, *lsn)) {
+                updates.push(Update {
+                    lsn: *lsn,
+                    timeline_id: timeline.timeline_id,
+                    command: Command::Update(*size),
+                });
+
+                needed_cache.insert((timeline.timeline_id, *lsn));
+            } else {
+                let timeline = Arc::clone(&timeline);
+                let parallel_size_calcs = Arc::clone(limit);
+                joinset.spawn(calculate_logical_size(parallel_size_calcs, timeline, *lsn));
```

Comment:
I'm not sure we have to fix it right now, but generally I'm against such approach: 
* every timeline could be decomposed as [initdb_lsn, ..., interesting_lsn_1, ..., interesting_lsn_i, ..., disk_consistent_lsn, ...] — only data after `disk_consistent_lsn` changes, the rest of the data is always constant, so it does not make any sense to recalculate the start of the timeline multiple times
* `interesting_lsns` consists of branch points, which are usually placed on top of some data due to nature of the branching: we create them to experiment on some data without spoiling it and I would imagine that users would create those branches close to their latest data, meaning that all those calculations will be heavy 
* Plans like https://neondb.slack.com/archives/C026T7K2YP9/p1667293260574949 could lead us to many-many branches created, which will slow the entire process down, since we have to wait for them all to be calculated

So, the whole process gets slower and slower the more branches we'll create of a timeline, even if timeline's data is not changing at all: that is very odd.
And then, there's an initial logical size computation background task, that's not even considered as part of these calculations, even though it does the same essentially.

Reply:
Overall this is a good point, and I had not noticed that discussion/idea.

Right now however this will only be provided via manually used http api so the danger side is low, however going back the original idea of a metric calculated periodically this size calculation has to become much more efficient overall, in addition to likely big fixes around where the data is actually gathered, it is likely to be wrong here (see how `refresh_gc_info` returns only the timelines which have grown past the gc horizon).

Reply:
Initial size computetation is now used as a cache as well, optimistically, in https://github.com/neondatabase/neon/pull/2714/commits/2fdef1aa5887fecebe14fbece64b930e8e14d3af if we decide to keep it. Other thread where this was discussed more: https://github.com/neondatabase/neon/pull/2714#discussion_r1011694080

Note that the initdb_lsn is unlikely to be used here however.

Related note for anyone coming spelunking after, in tests we noted that the act of branching increases the ancestor timeline lsn, moving the branch point away (in tests amount of gc_horizon -- might be the miniscule gc_horizon only) off where it was thought to be, leading to new size recalculation.